Unwind Attribute without Python Changes

jaisnan opened the pull request:
### Description of changes: 

Adds Unwind Attribute without Python Changes. The `kani-metadata.json` file now has the following structure

``` json
{
    "proof_harnesses": [
        {
            "pretty_name": "harness",
            "mangled_name": "harness",
            "original_file": "/home/ubuntu/kani/tests/cargo-kani/simple-unwind-annotation/src/main.rs",
            "original_line": "17",
            "unwind_value": 10
        },
        {
            "pretty_name": "harness_2",
            "mangled_name": "harness_2",
            "original_file": "/home/ubuntu/kani/tests/cargo-kani/simple-unwind-annotation/src/main.rs",
            "original_line": "26",
            "unwind_value": null
        }
    ]
}
```

Given that the rust library looks like -

``` rust
#[kani::proof]
#[kani::unwind(10)]
fn harness() {
...
}

#[kani::proof]
fn harness_2() {
...
}
```
Update - The PR has been updated to include handling for the following scenarios, all of which produce user errors which point to suggestions on how to fix the attribute.

``` rust
#[kani::proof]
#[kani::unwind(10,5)]
fn harness_3() {
...
}

#[kani::unwind(8)]
fn harness_4() {
...
}

#[kani::proof]
#[kani::proof]
fn harness_5() {
...
}

#[kani::proof(some, argument2)]
fn harness_6() {
...
}
```


### Resolved issues:

Resolves - partly https://github.com/model-checking/kani/issues/600 and https://github.com/model-checking/kani/issues/689.

### Call-outs:

The scripting needs to parse the metadata file and apply the unwind value to the target function.

### Testing:

* How is this change tested? Somewhat

* Is this a refactor change? No

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

tedinski commented on library/kani_macros/src/lib.rs:
```diff
@@ -49,3 +50,29 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+
+    // Add the string that looks like - #[kanitool::unwind_value_]
+    result.extend(insert_string.parse::<TokenStream>().unwrap());
+    // No mangle seems to be necessary as removing it prevents all the attributes in a lib from being read
```
I don't understand this comment. Unwind should be on something that already has `#[proof]` surely?

jaisnan replied:
This was a design choice made to separate unwind as an attribute from proof i.e, the user should be able to call unwind on harnesses without the proof attribute. This might be a disagreement in design that should be resolved before we add other attributes as well.

tedinski commented on library/kani_macros/src/lib.rs:
```diff
@@ -49,3 +50,29 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+pub fn unwind(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let mut result = TokenStream::new();
+
+    // Translate #[kani::unwind(arg)] to #[kanitool::unwind_arg_] for easier handling
```
Is this because you're not sure how to parse the attributes, or because you figured it out and it looked monstrous?

jaisnan replied:
I thought it created safer parsing conditions.

tedinski commented on src/kani-compiler/rustc_codegen_kani/src/context/metadata.rs:
```diff
@@ -19,8 +19,19 @@ pub struct HarnessMetadata {
 #[derive(Serialize)]
 pub struct KaniMetadata {
     pub proof_harnesses: Vec<HarnessMetadata>,
+    pub unwind_metadata: Vec<UnwindMetadata>,
```
IMO, the unwind value should be an `Option<u32>` on `HarnessMetadata`, not its own structure here.

jaisnan replied:
This too comes from a fork in the design i think.

adpaco-aws reviewed (changes requested):
Unwinding values should be part of the harness metadata and not a separate struct. For me it is okay to keep it a struct (not just a `u32` value) since long term we want to support unwinding values for individual loops in addition to this one.

jaisnan commented on library/kani_macros/src/lib.rs:
```diff
@@ -49,3 +50,29 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+#[cfg(not(kani))]
+#[proc_macro_attribute]
+pub fn unwind(_attr: TokenStream, _item: TokenStream) -> TokenStream {
+    // Not-RMC, Not-Test means this code shouldn't exist, return nothing.
```
My brain is still stuck on the old name sadly. Yes, i can do that , it might make the documentation clearer.

jaisnan commented on src/kani-compiler/rustc_codegen_kani/src/context/goto_ctx.rs:
```diff
@@ -59,6 +59,7 @@ pub struct GotocCtx<'tcx> {
     pub current_fn: Option<CurrentFnCtx<'tcx>>,
     pub type_map: FxHashMap<InternedString, Ty<'tcx>>,
     pub proof_harnesses: Vec<HarnessMetadata>,
+    pub unwind_metadata: Vec<UnwindMetadata>,
```
This was an active design choice that I think needs to be discussed. I was under the opinion that each attribute needs to be able to be called separately.

jaisnan commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -243,11 +245,33 @@ impl<'tcx> GotocCtx<'tcx> {
+                // In later Attributes, it will be easy to parse the regex for the attribute arguments, regardless
+                // of the type of the argument itself if we capture all the patterns as a vector of strings of the format
+                // [command, argument]
+                let captures = input_re.captures(attribute_string).map(|captures| {
```
It makes parsing convenient. I tried searching for better solutions to parsing the but I think this solves it the easiest way. Happy to listen to any alternate ways of doing this.

jaisnan commented:
Updated the PR as per suggestions.

adpaco-aws commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -243,16 +244,65 @@ impl<'tcx> GotocCtx<'tcx> {
+
+        // In the case when there's only one proof attribute (correct behavior), create harness and modify it
+        // depending on each subsequent attribute that's being called by the user.
+        if proof_attribute_vector.len() == 1 {
```
These cases do not seem to overlap so I am not sure why you are using if-then-else statements here. I would also recommend handling them in small functions if possible.

jaisnan replied:
They do overlap as it's necessary to check to how many proof attributes have been declared per harness and depending on the number, there's different handling that needs to happen. The entire function has a single purpose so it didn't seem necessary to factor it out into smaller functions.

celinval replied:
@jaisnan, if I understand this code correctly, you are trying to restrict any other `kani` attribute to things that are tagged with the `proof` attribute, right?

If that is the case, I would recommend to change this to have something like the following structure:
```rust
        let attrs: kani_attributes(self.tcx.get_attrs(instance.def_id()));  // BTreeMap<&str, &rustc_ast::Attribute>
        if attrs.contains_key("proof") {
            // Handle attributes here.
        } else {
            if attrs.len() > 0 {
                // Warning: We currently only support attributes on proof harnesses. The attribute {} will be ignored.
            }
        }
```

jaisnan replied:
I think using a BTreeMap would prevent us from checking if there are duplicate attributes in the same harness or not. If someone calls two `unwind` annotations then, the last one is going to be read and parsed if we use a BTreeMap I think. 

We ideally should not just check if a key exists but also how many instances of them. Maybe I can use a multimap? https://docs.rs/btreemultimap/latest/btreemultimap/ ?

adpaco-aws commented on src/kani-compiler/rustc_codegen_kani/src/context/metadata.rs:
```diff
@@ -17,6 +17,8 @@ pub struct HarnessMetadata {
     /// The line in that file where the proof harness begins
     pub original_line: String,
+    /// Optional data to store unwind value
+    pub unwind_value: Option<u128>,
```
I thought we agreed having a specific struct for unwind data where this would be an optional value. Also, `Option<u32>` should be enough.

jaisnan replied:
A struct didn't seem necessary for now as we're just targeting a default value for the entire harness. The struct and related changes will be added with the changes to map each loop to it's own harness.

celinval replied:
Maybe create a type alias here that we can extend later.

```rust
type UnwindMetadata = usize;
```

jaisnan replied:
do we want`usize` or `u32/64`. Doesnt `usize` tie the limit of the unwind value to the target architecture?

celinval replied:
I suggested `usize` because that's usually used to restrict structure sizes, but in  practice, I highly doubt someone is going to use a large unwind parameter. So `u32` is fine too.

tedinski replied:
IMO stay an integer. Any future unwinding information (e.g. unwindset) can be provided in separate fields. We're never going to change `unwind_value` because it's just the thing we pass to cbmc as `--unwind` and that's it.

tedinski replied:
I don't know why any future support for those wouldn't just add an `unwind_max_value` as well. There doesn't seem to be a value to nesting it. We control the producer (kani-compiler) and consumer (cargo-kani) of this structure, so if it's looks like a mistake later, we can just fix it. So simplest choice seems best to me.

But I don't feel strongly about it.

adpaco-aws commented on tests/cargo-kani/simple-unwind-annotation/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[dependencies]
+
+[kani.flags]
+output-format = "old"
```
No newline here yet.

jaisnan replied:
will add, thanks!

adpaco-aws commented on tests/cargo-kani/simple-unwind-annotation/src/main.rs:
```diff
@@ -0,0 +1,79 @@
+    }
+}
+
+// #[kani::proof]
```
Why is this commented code included? If there is a reason to, add a comment explaining why.

celinval replied:
Can we change these to be UI tests that we ensure that Kani prints the correct error message?

jaisnan replied:
sure

adpaco-aws commented on tests/kani/Unwind-Attribute/fixme_lib.rs:
```diff
@@ -0,0 +1,38 @@
+
+// rmc-flags: --no-unwinding-checks
+
+// Fix me
```
Please add the reason as to why this is a fix me test, including relevant issues if possible.

jaisnan commented on library/kani_macros/src/lib.rs:
```diff
@@ -49,3 +51,27 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
     // )
 }
+
+#[cfg(not(kani))]
+#[proc_macro_attribute]
+pub fn unwind(_attr: TokenStream, _item: TokenStream) -> TokenStream {
+    // This code is never called by default as the config value is set to kani
+    TokenStream::new()
+}
```
From what i understand, it's necessary to declare for when the configuration is set to anything but Kani (either deliberately or accidentally). in that case, we don't want the unwind attributes to cause a panic so we just return an empty tokenstream.

celinval commented on library/kani_macros/src/lib.rs:
```diff
@@ -36,9 +37,10 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+pub fn proof(attr: TokenStream, item: TokenStream) -> TokenStream {
     let mut result = TokenStream::new();
 
+    assert!(attr.to_string().len() == 0, "#[kani::proof] does not take any arguments");
```
Nice!

celinval commented on library/kani_macros/src/lib.rs:
```diff
@@ -49,3 +51,27 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+pub fn unwind(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let mut result = TokenStream::new();
+
+    // Translate #[kani::unwind(arg)] to #[kanitool::unwind(8)]
```
s/ #[kanitool::unwind(8)]/ #[kanitool::unwind(arg)]

celinval commented on library/kani_macros/src/lib.rs:
```diff
@@ -49,3 +51,27 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+
+#[cfg(kani)]
+#[proc_macro_attribute]
+pub fn unwind(attr: TokenStream, item: TokenStream) -> TokenStream {
```
Can you please document how this should be used? What kind of arguments does it accept and mention that today we restrict its usage to proof_harnesses only.

celinval commented on library/kani_macros/src/lib.rs:
```diff
@@ -49,3 +51,27 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    // Add the string that looks like - #[kanitool::unwindvalue]
+    result.extend(insert_string.parse::<TokenStream>().unwrap());
+    // No mangle seems to be necessary as removing it prevents all the attributes in a lib from being read
+    result.extend("#[no_mangle]".parse::<TokenStream>().unwrap());
```
Do we still need this?

jaisnan replied:
No, not anymore. let me update that.

jaisnan replied:
Update :  We do need this. Without it, all unwind annotations without `kani::proof` just get ignored and this prevents us from showing an error to the user. IMO, that would confuse the user as they would not know what the correct syntax for the annotation is. So, I think we should keep no_mangle on.

celinval replied:
No mangle has a side effect of making a function public and potentially causing name conflicts (E.g.: https://github.com/rust-lang/rust/issues/28179).

BTW, how did you test this behavior? Can you please make sure that the function you are annotating isn't dead code. Feel free to ping me offline.

celinval replied:
Talked offline. We'll remove this.

celinval commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -263,20 +313,64 @@ impl<'tcx> GotocCtx<'tcx> {
+    }
+
+    /// Unwind strings of the format 'unwind_x' and the mangled names are to be parsed for the value 'x'
+    fn handle_kanitool_unwind(&mut self, attr: &Attribute, harness: &mut HarnessMetadata) {
```
Since we are planning to eventually support unwind outside of the metadata, I would make this function independent from harness metadata. I.e.:

```rust
fn handle_unwind(&mut self, attrs: &BTreeMap<&str, &Attribute>) -> Option<UnwindMetadata> {
}
```

BTW, see my other comment about `UnwindMetadata` type.

tedinski replied:
Not sure I understand this comment, if we're putting unwind attributes elsewhere from proof harnesses (imo, still an 'if'), that's going to require a ton of design work, the choices of which we can't really anticipate. This approach seems simple. Trying to anticipate what we might do later is overly complex for the time being.

tedinski replied:
Oh now I see what you mean. I'm not sure if it is cleaner though. The need to give good error messages demands we spot duplicates and things like that. If we try to make this function "cleaner" then we end up pushing a lot of unwind-specific logic (like detecting duplicate unwind attributes) up to the calling function instead of here in the 'unwind' handler.

This organization lets that calling function have that nice, clean "loop over kanitool attributes, call handler function" and that's it.

celinval replied:
Anyway, not a biggie. Let's move on.

celinval commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -263,20 +313,64 @@ impl<'tcx> GotocCtx<'tcx> {
     match &attr.kind {
         ast::AttrKind::Normal(ast::AttrItem { path: ast::Path { segments, .. }, .. }, _)
-            if segments.len() == 2 && segments[0].ident.as_str() == "kanitool" =>
+            if (!segments.is_empty()) && segments[0].ident.as_str() == "kanitool" =>
```
+1

celinval commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -263,20 +313,64 @@ impl<'tcx> GotocCtx<'tcx> {
 }
 
 /// If the attribute is named `kanitool::name`, this extracts `name`
 fn kanitool_attr_name(attr: &ast::Attribute) -> Option<String> {
```
maybe this should return `Option<&str>` since we keep converting it to `&str`.

jaisnan replied:
Returning Option<&str> would return a dangling pointer as the original string would be destroyed once the function returns. We need to return an Owned String so I don't think we can replace the return type.

celinval replied:
Indeed, I didn't notice that you changed the code to join the segments. For the original code, the return value would have the same lifetime as the argument `attr`.

BTW, why did you change it? This code might actually be doing the wrong thing. I.e., let's say you have the following attribute:
```rust
#[kanitool::type::sub_type]
```
I believe that you are going to return the string `typesub_type`.

celinval replied:
Talked offline, this is no longer needed. Going to revert the change.

celinval commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -263,20 +313,64 @@ impl<'tcx> GotocCtx<'tcx> {
     }
 }
+
+/// Extracts the integer value argument from the unwind attribute
```
Can you please update the comments in this function? The function is generic but the comments are only related to unwind.

jaisnan replied:
Will do!

celinval commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -263,20 +313,64 @@ impl<'tcx> GotocCtx<'tcx> {
+/// Extracts the integer value argument from the unwind attribute
+fn extract_integer_argument(attr: &Attribute) -> Option<u128> {
+    // Vector of meta items , that contain metadata about the annotation
+    let attr_args = attr.meta().and_then(|x| x.meta_item_list().map(|x| x.to_vec()))?;
```
you can call `attr.meta_item_list()` directly.

jaisnan replied:
Thanks!

celinval commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -263,20 +313,64 @@ impl<'tcx> GotocCtx<'tcx> {
 /// If the attribute is named `kanitool::name`, this extracts `name`
 fn kanitool_attr_name(attr: &ast::Attribute) -> Option<String> {
     match &attr.kind {
         ast::AttrKind::Normal(ast::AttrItem { path: ast::Path { segments, .. }, .. }, _)
```
Since you are changing this function, can you please replace this logic by calling `attr.name_or_empty()`?

celinval reviewed (commented):
Thanks for doing this! I have a few suggestions to clean the code a bit. Let me know if you disagree with any of them. :)

jaisnan commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -243,16 +244,65 @@ impl<'tcx> GotocCtx<'tcx> {
+                format!("Please use '#kani[proof]' above the annotation {}", attribute_vector[0].0)
+                    .as_str(),
+            );
+        } else {
```
When neither proof or any other attributes are used. This function is called on all instances in `codegen_function` so raising an error might break a lot of cases. Hence, the empty handling.

jaisnan commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -243,16 +245,64 @@ impl<'tcx> GotocCtx<'tcx> {
+        // Vectors for storing instances of each attribute type,
+        // TODO: This can be modifed to use Enums when more options are provided
+        let mut attribute_vector = vec![];
+        let mut proof_attribute_vector = vec![];
```
We need to maintain a count of the instances apart from the attributes themselves, hence vectors.

celinval commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -243,16 +245,64 @@ impl<'tcx> GotocCtx<'tcx> {
+        else if proof_attribute_vector.len() > 1 {
+            self.tcx
+                .sess
+                .span_err(proof_attribute_vector[0].span, "Only one Proof Attribute allowed");
```
Reduce this to warning.

celinval commented on src/kani-compiler/rustc_codegen_kani/src/codegen/function.rs:
```diff
@@ -263,20 +313,68 @@ impl<'tcx> GotocCtx<'tcx> {
+            _ => None,
+        }
+    }
+    // Return none if there are no unwind attributes or if there's too many unwind attributes
```
unwind comment in the generic function.

celinval reviewed (commented):
Here are some notes from our chat.

jaisnan commented:
1. Removed `no_mangle`
2. Refactored the parsing into functions, added utility functions
3. Added documentation for all user cases
4. Added a test to check failure of harness assert instead of `main`.

tedinski commented on library/kani_macros/src/lib.rs:
```diff
@@ -9,6 +9,7 @@
 //   RUSTFLAGS="-Zcrate-attr=feature(register_tool) -Zcrate-attr=register_tool(kanitool)"
 
 // proc_macro::quote is nightly-only, so we'll cobble things together instead
+extern crate proc_macro;
```
was this necessary?

tedinski commented on library/kani_macros/src/lib.rs:
```diff
@@ -49,3 +51,27 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    TokenStream::new()
+}
+
+// The attribute '#[kani::unwind(arg)]' can only be called alongside '#[kani::proof]'.
+// arg - Takes in a integer value (u32) that represents the unwind value for the harness.
+// Usage is restricted to proof harnesses only currently.
```
Doc-comments in Rust use 3 `/`s not 2, check all your function comments for this.

tedinski commented on src/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -241,18 +243,54 @@ impl<'tcx> GotocCtx<'tcx> {
     /// i.e. `#[kani::proof]` (which kani_macros translates to `#[kanitool::proof]` for us to handle here)
     fn handle_kanitool_attributes(&mut self) {
-        let instance = self.current_fn().instance();
+        let attributes_list = self.tcx.get_attrs(self.current_fn().instance().def_id());
+        let (proof_attribute_vector, attribute_vector) = create_attribute_vectors(attributes_list);
```
Naming suggestion: don't put the type into the name really.

IMO, these names are clearer as "all_attributes" "proof_attributes" "other_attributes".

tedinski commented on src/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -241,18 +243,54 @@ impl<'tcx> GotocCtx<'tcx> {
-                Some("proof") => self.handle_kanitool_proof(),
-                _ => {}
+        // Handle the attributes only if there is one proof attribute
+        if proof_attribute_count == 1 {
+            self.create_proof_harness(attribute_vector);
+        }
```
There's a very common programming convention to write functions like:

```
fn foo() {
  if error condition {
    // bail, whatever
  }
  if other error {
    // bail
  }
  // etc
  // no errors:
  the conventional, expected case.
}
```

So try that. Spot errors, raise them and return. Then just unconditionally `create_proof_harness` at the end of the function (no need for any empty else). By following the convention, functions are usually a lot easier to read (if only because people expect it)

jaisnan replied:
For the no-op case, can we now early return with the empty parenthesis?

tedinski commented on src/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -241,18 +243,54 @@ impl<'tcx> GotocCtx<'tcx> {
     }
 
     /// Update `self` (the goto context) to add the current function as a listed proof harness
-    fn handle_kanitool_proof(&mut self) {
+    fn handle_kanitool_proof(&mut self) -> HarnessMetadata {
```
Function doc comment needs updating. Also, we should rename this, maybe. "default_kanitool_proof" maybe?

tedinski commented on src/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -263,20 +301,93 @@ impl<'tcx> GotocCtx<'tcx> {
+
+    /// Unwind strings of the format 'unwind(x)' and the mangled names are to be parsed for the value 'x'
+    fn handle_kanitool_unwind(&mut self, attr: &Attribute, harness: &mut HarnessMetadata) {
+        // Check if some unwind value doesnt already exist
+        if harness.unwind_value.is_none() {
```
Same suggestion as previous comment on handling errors, then the expected case.

jaisnan replied:
Going to wrap the behavior with a match statement.

tedinski commented on src/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -263,20 +301,93 @@ impl<'tcx> GotocCtx<'tcx> {
+            // Get Attribute value and if it's not none, assign it to the metadata
+            if let Some(integer_value) = extract_integer_argument(attr) {
+                // Convert the extracted u128 and convert to u32
+                assert!(
+                    integer_value < u32::MAX.into(),
+                    "Value above maximum permitted value - u32::MAX"
+                );
```
Good call. But we might want to `span_err` instead of assert.

tedinski commented on src/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -263,20 +301,93 @@ impl<'tcx> GotocCtx<'tcx> {
+
+// Create two vectors, proof_vector and attribute_vector which seperates the list of attributes into
+// proof and non-proof for easier parsing
+fn create_attribute_vectors(
```
Rename `partition_kanitool_attributes` maybe?

Second suggestion: try improving the doc-comment enough that all the comments in the body aren't necessary anymore. This function is small enough this is possible and would very likely be an improvement.

tedinski commented on src/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -241,18 +243,54 @@ impl<'tcx> GotocCtx<'tcx> {
+        // In the case when there's only one proof attribute (correct behavior), create harness and modify it
+        // depending on each subsequent attribute that's being called by the user.
+        let mut harness_metadata = self.handle_kanitool_proof();
+        if attribute_vector.len() > 0 {
```
unnecessary if

tedinski commented on src/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -241,18 +243,54 @@ impl<'tcx> GotocCtx<'tcx> {
+                    "unwind" => {
+                        self.handle_kanitool_unwind(attribute_tuple.1, &mut harness_metadata)
+                    }
+                    _ => {}
```
maybe we could error or warn?

tedinski reviewed (commented):
Almost there!

tedinski commented on tests/ui/unsupported-annotation/main.rs:
```diff
@@ -0,0 +1,21 @@
+}
+
+#[kani::proof]
+#[kani::test_annotation]
```
FYI: this isn't testing your code, you'd have to write `kanitool::` for that

tedinski commented on src/kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -241,18 +243,54 @@ impl<'tcx> GotocCtx<'tcx> {
+        if proof_attributes.is_empty() && !other_attributes.is_empty() {
+            self.tcx.sess.span_err(
+                other_attributes[0].1.span,
+                "Please use '#kani[proof]' above the annotation",
```
typo

tedinski commented on library/kani_macros/src/lib.rs:
```diff
@@ -49,3 +50,26 @@ pub fn proof(_attr: TokenStream, item: TokenStream) -> TokenStream {
+#[proc_macro_attribute]
+pub fn unwind(_attr: TokenStream, _item: TokenStream) -> TokenStream {
+    // When the config is not kani, we should leave the function alone
+    _item
```
Does it issue a warning or anything about the _? (That usually indicated unused, so maybe this should just be `item` not `_item`)

tedinski reviewed (approved):
Assuming CI passes, looks largely good. Maybe one last typo.
